hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/S3Acl.java

```diff
@@ -0,0 +1,388 @@
+    ACLType(String val) {
+      value = val;
+    }
+
+
+    public static ACLType getType(String typeStr) {
+      for(ACLType type: ACLType.values()) {
+        if (type.getValue().equals(typeStr)) {
+          return type;
+        }
+      }
+      return null;
+    }
+  }
+
+  enum ACLIdentityType {
+    USER("CanonicalUser", true, "id"),
+    GROUP("Group", false, "url"),
+    USER_BY_EMAIL("AmazonCustomerByEmail", false, "emailAddress");
+
+    public String getGranteeType() {
+      return granteeType;
+    }
+
+    public String getHeaderType() {
+      return granteeInHeader;
+    }
+
+    /**
+     *  Grantee type in body XML.
+     */
+    private final String granteeType;
+
+    /**
+     * Is this type supported or not.
+     */
+    private final boolean supported;
+
+    /**
+     * Grantee type in header.
+     */
+    private final String granteeInHeader;
+
+    /**
+     * Init OzoneACLtypes enum.
+     *
+     * @param val String type for this enum.
+     */
+    ACLIdentityType(String val, boolean support, String headerType) {
+      granteeType = val;
+      supported = support;
+      granteeInHeader = headerType;
+    }
+
+    boolean isSupported() {
+      return supported;
+    }
+
+    public static ACLIdentityType getTypeFromGranteeType(String typeStr) {
+      for(ACLIdentityType type: ACLIdentityType.values()) {
+        if (type.getGranteeType().equals(typeStr)) {
+          return type;
+        }
+      }
+      return null;
+    }
+
+    public static ACLIdentityType getTypeFromHeaderType(String typeStr) {
+      for(ACLIdentityType type: ACLIdentityType.values()) {
+        if (type.getHeaderType().equals(typeStr)) {
+          return type;
+        }
+      }
+      return null;
+    }
+  }
+
+  public static boolean isGranteeTypeSupported(String typeStr) {
+    ACLIdentityType type =  ACLIdentityType.getTypeFromGranteeType(typeStr);
+    return type == null ? false : type.isSupported();
+  }
+
+  public static boolean isHeaderTypeSupported(String typeStr) {
+    ACLIdentityType type =  ACLIdentityType.getTypeFromHeaderType(typeStr);
+    return type == null ? false : type.isSupported();
+  }
+
+  public static List<Grant> ozoneNativeAclToS3Acl(OzoneAcl ozoneAcl) {
+    // Since currently only "CanonicalUser" is supported, which maps to Ozone
+    // "USER"
+    List<Grant> grantList = new ArrayList<>();
+    if (ozoneAcl.getType() != IAccessAuthorizer.ACLIdentityType.USER) {
+      return grantList;
+    }
+
+    Grantee grantee = new Grantee();
+    grantee.setDisplayName(ozoneAcl.getName());
+    grantee.setId(ozoneAcl.getName());
+
+    List<IAccessAuthorizer.ACLType> acls = ozoneAcl.getAclList();
+    if (acls.contains(IAccessAuthorizer.ACLType.ALL)) {
+      Grant grant = new Grant();
+      grant.setGrantee(grantee);
+      grant.setPermission(ACLType.FULL_CONTROL.toString());
+      grantList.add(grant);
+      return grantList;
+    } else if (acls.contains(IAccessAuthorizer.ACLType.WRITE_ACL)) {
```

Comment:
should we allow other combinations here? such as READ+WRITE?  I feel if we remove "else" here, the 1-1 mapping between Ozone AclType to the S3 ACLType can be added.

Reply:
If I understand your question correctly,  when else is removed, then we will have a grant list with both FULL_CONTROL and WRITE_ACL grant for the same user, which is kind of information duplicate.